[AUDIO_WORKLET] Fix emscripten_audio_worklet_post_function_sig

mighty1231 opened the pull request:
It fixes emscripten_audio_worklet_post_function_sig, which was previously not worked, and enhances test to cover all
emscripten_audio_worklet_post_function_*.

In order to fix, it changes the old term `readAsmConstArgs` to the new term `readEmAsmArgs`, due to conflict of https://github.com/emscripten-core/emscripten/pull/16449 and https://github.com/emscripten-core/emscripten/pull/18218

Background history:

1. #16449 proposed, introducing AudioWorklet earlier
2. #18218 proposed, renaming readAsmConstArgs to readEmAsmArgs
3. #18218 merged earlier
4. #16449 merged later, using the old term `readAsmConstArgs`

mighty1231 commented:
I see test job 'test-browser-chrome-wasm64-4gb' has failed from `test_offset_converter_pthread`, it seems not directly related to this PR. 

```ansi
  test_offset_converter_pthread (test_browser.browser64_4gb) ... cache:INFO: [32mgenerating system asset: symbol_lists/521c79b8fb310ffb91dc32c4132585ef16cab6fd.json... (this will be cached in "/root/cache/symbol_lists/521c79b8fb310ffb91dc32c4132585ef16cab6fd.json" for subsequent builds)[0m
cache:INFO: [32m - ok[0m
[unresponsive tests: 1]
common:INFO: [32mRestarting browser process[0m
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 42318)
Traceback (most recent call last):
  File "/usr/lib/python3.8/socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.8/socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.8/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.8/http/server.py", line 647, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib/python3.8/socketserver.py", line 747, in __init__
    self.handle()
  File "/usr/lib/python3.8/http/server.py", line 427, in handle
    self.handle_one_request()
  File "/usr/lib/python3.8/http/server.py", line 415, in handle_one_request
    method()
  File "/root/project/test/common.py", line 2186, in do_GET
    self.wfile.write(b'(wait)')
  File "/usr/lib/python3.8/socketserver.py", line 826, in write
    self._sock.sendall(b)
BrokenPipeError: [Errno 32] Broken pipe
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 42328)
Traceback (most recent call last):
  File "/usr/lib/python3.8/socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.8/socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.8/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.8/http/server.py", line 647, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib/python3.8/socketserver.py", line 747, in __init__
    self.handle()
  File "/usr/lib/python3.8/http/server.py", line 427, in handle
    self.handle_one_request()
  File "/usr/lib/python3.8/http/server.py", line 415, in handle_one_request
    method()
  File "/root/project/test/common.py", line 2156, in do_GET
    self.wfile.write(b'OK')
  File "/usr/lib/python3.8/socketserver.py", line 826, in write
    self._sock.sendall(b)
BrokenPipeError: [Errno 32] Broken pipe
----------------------------------------
common:INFO: [32mLaunching browser: ['/usr/bin/google-chrome', '--no-first-run', '-start-maximized', '--no-sandbox', '--use-gl=swiftshader', '--user-data-dir=/tmp/chrome-emscripten-profile', '--enable-experimental-web-platform-features', '--headless=new', '--remote-debugging-port=1234', '--enable-experimental-webassembly-features', '--js-flags=--experimental-wasm-stack-switching --experimental-wasm-type-reflection', '--disk-cache-dir=/dev/null', '--disk-cache-size=1', '--media-cache-size=1', '--disable-application-cache', '--incognito'][0m
[test error (see below), automatically retrying]
Expected to find '/report_result?exit:0
' in '[no http server activity]
', diff:
```

sbc100 commented:
test_offset_converter_pthread is a known flaky test

sbc100 commented on test/webaudio/audioworklet_post_function.c:
```diff
@@ -21,7 +21,7 @@ void MessageReceivedInAudioWorkletThread(int a, int b) {
   assert(emscripten_current_thread_is_audio_worklet());
   assert(a == 42 && b == 9000);
-  emscripten_audio_worklet_post_function_viii(EMSCRIPTEN_AUDIO_MAIN_THREAD, MessageReceivedOnMainThread, /*d=*/1, /*e=*/2, /*f=*/3);
+  emscripten_audio_worklet_post_function_sig(EMSCRIPTEN_AUDIO_MAIN_THREAD, (void *)MessageReceivedOnMainThread, "iii", /*d=*/1, /*e=*/2, /*f=*/3);
```
Should we not be tests both the `_viii` form and the `_sig` form?

So IIUC that `emscripten_audio_worklet_post_function_sig` simply does not work today and is not tested?

(That makes me wonder if its really needed as part of the API?)

mighty1231 replied:
`emscripten_audio_worklet_post_function_sig` is not currently tested in the test suite, and it does not work properly at the moment. So that I changed existing test case. I could enhance the test, but I’d like to hear your thoughts on whether this version of test case is truly necessary. What do you think?

There are 2 threads for test: `main` and `audio_worklet_thread`

**before this PR**

1. `main` --> `audio_worklet_thread`, calling `_vii`
2. `audio_worklet_thread` --> `main`, calling `_viii`

**after this PR**

1. `main` --> `audio_worklet_thread`, calling `_vii`
2. `audio_worklet_thread` --> `main`, calling `_sig`

**possible test case (necessary?)**

1. `main` --> `audio_worklet_thread`, calling `_vii`
2. `audio_worklet_thread` --> `main`, calling `_sig`
3. `main` --> `audio_worklet_thread`, calling `_sig`
4. `audio_worklet_thread` --> `main`, calling `_viii`

sbc100 replied:
As long as at least one of the `_vxxx` functions is tested that seems reasonable.   Having coverage for *all* difference `_vxxx` functions seems even better.

mighty1231 replied:
I enhanced the test case with a new commit, referring other test case - `test/wasm_worker/post_function.c`.

sbc100 commented:
Can you mention the PR description the name of the function (`emscripten_audio_worklet_post_function_sig`) and mention that it has never worked (for the reasons you describe) and was not tested previously?

mighty1231 commented:
> Can you mention the PR description the name of the function (`emscripten_audio_worklet_post_function_sig`) and mention that it has never worked (for the reasons you describe) and was not tested previously?

I changed the PR description.

Previously, I set PR description as same as my commit message. Now, PR description differs to my commit message. Is it OK?

```diff
+It fixes emscripten_audio_worklet_post_function_sig, which was previously not worked, and changes existing test to check it.
```

mighty1231 commented:
> test_offset_converter_pthread is a known flaky test

How about changing `test_offset_converter.c` as follow?

```diff
void magic_test_function(void) {
  int result = EM_ASM_INT({
    function report(x) {
      out(x);
      fetch('http://localhost:8888?stdout=' + encodeURIComponent(x));
    }
-    report('magic_test_function: input=' + $0);
-    var converted = wasmOffsetConverter.getName($0);
-    report('magic_test_function: converted=' + converted);
-    return converted == 'magic_test_function';
+    try {
+      report('magic_test_function: input=' + $0);
+      var converted = wasmOffsetConverter.getName($0);
+      report('magic_test_function: converted=' + converted);
+      return converted == 'magic_test_function';
+    } catch (e) {
+      report(e?.message + e?.stack);
+      return false;
+    }
  }, get_pc());
  assert(result);
}
```

mighty1231 commented:
Now test `test_pthread_wait64_notify` not works. Where does the number `-43.357421875` comes from?

```
@@ -1,5 +1,8 @@
 main: creating worker
 main: entering timeout loop to wait for wasm worker to run
+(node:27876) TimeoutNegativeWarning: -43.357421875 is a negative number.
+Timeout duration was set to 1.
+(Use `node --trace-warnings ...` to show where the warning was created)
 worker_main
 Waiting on address with unexpected value should return 'not-equal'
 Waiting on address with unexpected value should return 'not-equal' also if timeout==0
```

sbc100 commented on test/webaudio/audioworklet_post_function.c:
```diff
@@ -6,28 +6,121 @@
+//   1. post function from main thread to audio worklet thread, with testingAudioWorklet = true
+//   2. post function from audio worklet thread to main thread, with testingAudioWorklet = false
+volatile bool testingAudioWorklet = true;
+volatile int success = 0;
```
Use `_Atomic` rather than `volatile`?   Since that is better defined for multi-threaded access I believe.

mighty1231 replied:
I agree with you

sbc100 commented on test/webaudio/audioworklet_post_function.c:
```diff
@@ -6,28 +6,121 @@
+//   1. post function from main thread to audio worklet thread, with testingAudioWorklet = true
+//   2. post function from audio worklet thread to main thread, with testingAudioWorklet = false
+volatile bool testingAudioWorklet = true;
+volatile int success = 0;
```
How about calling this something like `callback_count`?

mighty1231 replied:
I recognized ambiguity. However, I used `callbackCount` since many tests seems using camel case, instead snake case.

sbc100 replied:
sgtm

sbc100 commented on test/webaudio/audioworklet_post_function.c:
```diff
@@ -6,28 +6,121 @@
+    emscripten_audio_worklet_post_function_vdd(EMSCRIPTEN_AUDIO_MAIN_THREAD, vdd, 2.5, 3.5);
+    emscripten_audio_worklet_post_function_vddd(EMSCRIPTEN_AUDIO_MAIN_THREAD, vddd, 4.5, 5.5, 6.5);
+    emscripten_audio_worklet_post_function_sig(EMSCRIPTEN_AUDIO_MAIN_THREAD, viiiiiidddddd, "iiiiiidddddd", 10, 11, 12, 13, 14, 15, 16.5, 17.5, 18.5, 19.5, 20.5, 21.5);
+    emscripten_audio_worklet_post_function_v(EMSCRIPTEN_AUDIO_MAIN_THREAD, sided_test_finished);
```
Why not use a different function here?  e.g. `do_exit`?    Then you don't need the `testingAudioWorklet` global maybe?

mighty1231 replied:
1. On resolving conflict with merged PR #23695, I used `emscripten_force_exit(0);` as the PR used, instead of `do_exit`.
2. Originally, `testingAudioWorklet` is on the purpose to check `emscripten_current_thread_is_audio_worklet()` behaves correctly on each callbacks. I changed the name of it to `onAudioWorkletThread`, for legibility.

sbc100 replied:
Sure, it just seems a little odd that this callback has two completely different behaviors depending on the thread.   Why not have two different callbacks, one which gets called on the worklet and triggers the second set of callbacks, and one which run on the main thread call `on_exit` which just does the exiting?

mighty1231 replied:
I agree

sbc100 commented on test/webaudio/audioworklet_post_function.c:
```diff
@@ -6,28 +6,121 @@
+  } else { // for the second call, finishing checks for audio worklet -> main
+#ifdef REPORT_RESULT
+    REPORT_RESULT(success);
+#endif
```
Can we just access `success == XX` here and return zero?

mighty1231 replied:
As of the previous review, during resolving conflict, I changed `emscripten_force_exit(0);`, and added assertion to check `callbackCount` with 16=8 for `main->awthread` + 8 for `awthread->main`

mighty1231 commented:
1. Last force pushed commit was just fixing typo, due to changing all of `success` to `callbackCount`.

   ```diff
   -void WebAudioWorkletThreadInitialized(EMSCRIPTEN_WEBAUDIO_T audioContext, bool callbackCount, void *userData) {
   +void WebAudioWorkletThreadInitialized(EMSCRIPTEN_WEBAUDIO_T audioContext, bool success, void *userData) {
   ```

2. I added just *experimental* commit, in order to try to resolve test flakiness, but I have not changed PR title and description. `Edit this PR` vs `Creating new PR`?

sbc100 commented:
If you are able to help debug the flakiness that would be amazing, but that should part of a separate PR

mighty1231 commented:
> If you are able to help debug the flakiness that would be amazing, but that should part of a separate PR

@sbc100 I separated PR with #23716

mighty1231 commented:
[Is `test-browser.test_codesize_hello_dylink` flaky?](https://circleci.com/api/v1.1/project/github/emscripten-core/emscripten/903213/output/111/0?file=true&allocation-id=67b820e2452fa659160153b0-0-build%2FABCDEFGH)

```
======================================================================
FAIL [0.003s]: test_codesize_hello_dylink (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/root/project/test/common.py", line 897, in resulting_test
    return func(self, *args)
  File "/root/project/test/test_other.py", line 9127, in test_codesize_hello
    self.run_codesize_test('hello_world.c', *args)
  File "/root/project/test/test_other.py", line 9017, in run_codesize_test
    self.check_expected_size_in_file('wasm', size_file, wasm_size)
  File "/root/project/test/common.py", line 1679, in check_expected_size_in_file
    self.assertLess(ratio, size_slack)
  File "/usr/lib/python3.8/unittest/case.py", line 1298, in assertLess
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: 0.15836079077429985 not less than 0.05

```

[Is `test-browser-firefox.test_wasm_worker_lock_async_acquire_minimal_runtime` flaky?](https://circleci.com/api/v1.1/project/github/emscripten-core/emscripten/903222/output/109/0?file=true&allocation-id=67b820e2b173e3160c9b3bac-0-build%2FABCDEFGH)

```
  test_wasm_worker_lock_async_acquire_minimal_runtime (test_browser.browser) ... cache:INFO: [32mgenerating system asset: symbol_lists/6b6f462a54bb66fcbe3f154a623efbd87bf62831.json... (this will be cached in "/root/cache/symbol_lists/6b6f462a54bb66fcbe3f154a623efbd87bf62831.json" for subsequent builds)[0m
cache:INFO: [32m - ok[0m
[test error (see below), automatically retrying]
Expected to find '/report_result?exit:0
' in '//report_result?exception:can't access property "postMessage", U[a] is undefined / emscripten_wasm_worker_post_function_v@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597:241:3
@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597 line 275 > WebAssembly.instantiate:wasm-function[22]:0x854
@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597 line 275 > WebAssembly.instantiate:wasm-function[23]:0x9ed
emscripten_lock_async_acquire/d/<@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597:220:11
', diff:

--- expected
+++ actual
@@ -1 +1,5 @@
-/report_result?exit:0
+//report_result?exception:can't access property "postMessage", U[a] is undefined / emscripten_wasm_worker_post_function_v@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597:241:3
+@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597 line 275 > WebAssembly.instantiate:wasm-function[22]:0x854
+@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597 line 275 > WebAssembly.instantiate:wasm-function[23]:0x9ed
+emscripten_lock_async_acquire/d/<@blob:http://localhost:8888/623e1592-52e3-4247-aca1-8ed10e372597:220:11
+
```

cwoffenden commented:
Just to say nice work, I think every function of the AW API now has code coverage with this addition and #23659.

sbc100 commented:
The codesize test I believe just needs to be rebaselines.  I'll that now.

sbc100 commented:
Rebaselined: fc15522a1
